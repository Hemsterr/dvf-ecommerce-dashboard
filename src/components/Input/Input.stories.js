// Libs
import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

// Components
import Input from './index'

storiesOf('Input', module)
  .add('Default', () => <Input />)
  .add('Input primary', () => <Input label="TextBox" isBorder />)
  .add('Input has a error message', () => (
    <Input label="TextBox" errorMessage="You're enterd wrong format" isBorder />
  ))
  .add('Input has handle event on blur', () => (
    <Input label="TextBox" isBorder handleOnBlur={action('blurred')} />
  ))
  .add('Input has handle event on enter', () => (
    <Input label="TextBox" handleOnKeyDown={action('entered')} isBorder />
  ))
