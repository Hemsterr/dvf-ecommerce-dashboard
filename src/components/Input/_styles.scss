.input-group {
  display: flex;

  &__without-border {
    border: none;
  }

  &__label, &__label-error {
    @include text-size(pxToRem(12px), pxToRem(2px), pxToRem(40px));
    font-family: $font-base-semi-bold;
  }

  &__label {
    color: $black;
    padding-right: pxToRem(14px);
  }

  &__label-error {
    color: $red;
    text-transform: uppercase;
  }

  &__text {
    width: calc(100% - 30px);
    padding: 0 pxToRem(15px);
    height: pxToRem(40px);
    font-family: $font-base-regular;
    color: $black;
    text-align: center;
    @include text-size(pxToRem(12px), pxToRem(4.5px), pxToRem(20px));
    background-color: $white;
    text-transform: uppercase;

    &:-ms-input-placeholder { /* IE10â€“11 */
      color: $light-gray;
    }

    &:focus {
      outline: none;
    }
  }

  &__border {
    border: pxToRem(1px) solid $light-gray;
    height: pxToRem(28px);
    text-align: left;

    &--error {
      border-color: $red;
    }
  }

  &__tooltip {
    width: pxToRem(15px);
    height: pxToRem(15px);
    background: url('../../assets/icons/question-icon.png');
    background-size: cover;
    display: block;
    margin-left: pxToRem(8px);
  }
}

@media only screen and (max-width: $screen-xs) {
  .input-group {
    &__label-error {
      font-size: pxToRem(8px);
      line-height: pxToRem(20px);
    }
  }
}
