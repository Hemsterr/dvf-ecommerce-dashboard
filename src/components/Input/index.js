// @flow
import React from 'react'

// Constants
import keycodes from '../../constants/keyCodes'

type Props = {
  type: string,
  label: string,
  defaultValue: string,
  placeholder: string,
  errorMessage: string,
  keyDown: number,
  handleOnBlur: Function,
  handleOnKeyDown: Function,
  isBorder: boolean,
  className: string,
  name?: string,
  disabled?: boolean,
}

type State = {
  valueInput: string,
}

class Input extends React.PureComponent<Props, State> {
  static defaultProps = {
    type: 'text',
    label: '',
    defaultValue: '',
    placeholder: '',
    errorMessage: '',
    keyDown: keycodes.KEY_ENTER,
    handleOnBlur: () => {},
    handleOnKeyDown: () => {},
    isBorder: false,
    className: '',
    disabled: false,
  }

  constructor(props: Props) {
    super(props)
    const { defaultValue } = this.props

    this.state = {
      valueInput: defaultValue,
    }
  }

  // handle event on enter
  handleKeyDown = (event: Object) => {
    const { keyCode, target } = event
    const valueFromKeyDown = target.value
    const { handleOnKeyDown, keyDown } = this.props

    if (keyCode === keyDown) {
      handleOnKeyDown(valueFromKeyDown)
    }
  }

  render() {
    const {
      type,
      label,
      placeholder,
      errorMessage,
      isBorder,
      className,
      name,
      handleOnBlur,
      disabled,
    } = this.props
    const { valueInput } = this.state
    const classError = errorMessage ? 'input-group__text--error' : ''

    return (
      <div className={`${className} input-group`}>
        {label && <label className="input-group__label">{label}</label>}
        <div>
          <input
            type={type}
            className={`input-group__text ${classError} ${
              isBorder ? 'input-group__border' : 'input-group__without-border'
            }`}
            value={valueInput}
            name={name}
            placeholder={placeholder}
            onChange={event => this.setState({ valueInput: event.target.value })}
            onBlur={event => handleOnBlur(event.target.value.trim(), event.target.name)}
            onKeyDown={this.handleKeyDown}
            disabled={disabled}
          />
          {errorMessage && (
            <i className="input-group__label-error">{errorMessage}</i>
          )}
        </div>
      </div>
    )
  }
}

Input.defaultProps = {
  type: 'text',
  label: '',
  defaultValue: '',
  placeholder: '',
  errorMessage: '',
  keyDown: keycodes.KEY_ENTER,
  handleOnBlur: () => {},
  handleOnKeyDown: () => {},
  isBorder: false,
  className: '',
  disabled: false,
}

export default Input
