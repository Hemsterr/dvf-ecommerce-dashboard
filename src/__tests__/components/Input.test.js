// Libs
import React from 'react'
import { shallowToJson } from 'enzyme-to-json'
import { shallow } from 'enzyme'

// Component
import Input from '../../components/Input'

// Constants
import keycodes from '../../constants/keyCodes'

const mockBlurFunc = jest.fn()
const mockKeyDownFunc = jest.fn()
const mockProps = {
  type: 'text',
  label: 'User Name',
  defaultValue: 'Robert Downey Jr',
  placeholder: 'Enter on textbox, please!',
  errorMessage: 'You entered format wrong for User Name field',
  handleOnBlur: mockBlurFunc,
  handleOnKeyDown: mockKeyDownFunc,
  className: 'userName',
}

const wrapper = (props = {}) => {
  const component = shallow(<Input {...props} />)

  return component
}

describe('Input component', () => {
  const wrapperWithoutProps = wrapper()

  // Snapshot
  test('should render correctly Input component default props', () => {
    expect(shallowToJson(wrapperWithoutProps)).toMatchSnapshot()
  })

  // Testing props
  describe('Handle with props', () => {
    test('should is rendered default props', () => {
      expect(wrapperWithoutProps.find('input').props().defaultValue).toBeFalsy()
      expect(Input.defaultProps.handleOnBlur()).toBeUndefined()
      expect(Input.defaultProps.handleOnKeyDown()).toBeUndefined()
    })

    test('should is rendered with a custom Props', () => {
      const wrapperWithProps = wrapper(mockProps)

      // Expected with a placeholder
      expect(wrapperWithProps.find('input').props().placeholder).toMatch(
        mockProps.placeholder
      )

      // Expected with a defaultValue
      expect(wrapperWithProps.find('input').props().value).toMatch(
        mockProps.defaultValue
      )

      // Expected with a className
      expect(wrapperWithProps.find('.input-group').props().className).toMatch(
        mockProps.className
      )

      // Expected have a label
      expect(wrapperWithProps.find('label').props().children).toMatch(
        mockProps.label
      )

      // Expected have not label
      expect(wrapperWithoutProps.find('label')).toHaveLength(0)

      expect(wrapperWithoutProps.find('.userName')).toHaveLength(0)

      // Expected with a error message
      expect(
        wrapperWithProps.find('.input-group__label-error').props().children
      ).toMatch(mockProps.errorMessage)

      // Expected don't appear a error message
      expect(wrapperWithoutProps.find('.input-group__error')).toHaveLength(0)

      // Expected have contain class to change border color when input appear error
      expect(
        wrapperWithProps.find('input').hasClass('input-group__text--error')
      ).toEqual(true)

      // Expected don't contain class to change border color when input's not appear error
      expect(
        wrapperWithoutProps.find('input').hasClass('input-group__text--error')
      ).toEqual(false)
    })
  })

  describe('Handle with isPrimary props', () => {
    const InputComponent = wrapper({ ...mockProps, isBorder: true })
    expect(InputComponent.find('.input-group__border')).toHaveLength(1)
  })

  // Testing events
  describe('Checking events', () => {
    test('check the on Blur callback ', () => {
      const InputComponent = wrapper(mockProps)

      const input = InputComponent.find('input')
      const value = 'John Hart'
      const name = 'userName'

      input.simulate('blur', {
        target: { value, name },
      })
      expect(mockBlurFunc).toHaveBeenCalledWith(value, name)
      expect(InputComponent.find('.input-group__without-border')).toHaveLength(
        1
      )
      expect(InputComponent.find('.userName')).toHaveLength(1)
    })

    test('check the on Enter Press callback ', () => {
      const InputComponent = wrapper(mockProps)

      const input = InputComponent.find('input')
      const value = 'John Hart J'

      input.simulate('keydown', {
        keyCode: keycodes.KEY_ENTER,
        target: { value },
      })

      expect(mockKeyDownFunc).toHaveBeenCalledTimes(1)
    })

    test('check the on Change callback ', () => {
      const input = wrapperWithoutProps.find('input')
      const value = 'John Hart Jr'

      input.simulate('change', {
        target: { value },
      })

      expect(wrapperWithoutProps.find('input').props().value).toEqual(value)
    })
  })
})
