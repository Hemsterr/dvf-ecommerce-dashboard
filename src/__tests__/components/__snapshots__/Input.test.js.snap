// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render correctly Input component default props 1`] = `
<div
  className=" input-group"
>
  <div>
    <input
      className="input-group__text  input-group__without-border"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>
  <span
    className="input-group__tooltip"
    data-event="click focus"
    data-for="tooltip-guide"
  />
  <ReactTooltip
    border={true}
    clickable={false}
    effect="solid"
    globalEventOff="click"
    id="tooltip-guide"
    insecure={true}
    place="right"
    resizeHide={true}
    type="light"
    wrapper="div"
  >
    <div
      className="landing__tooltip"
    >
      <p>
        need help? 
        <span>
          How to measure guide
        </span>
      </p>
    </div>
  </ReactTooltip>
</div>
`;
